### PR TITLE
Rework structure of workflow store and metadata, provides roughly 5-f…

### PR DESCRIPTION
**Pull request by geoffjentry**
…old performance increase under load prior to DB gumming up. 

Things to note:
- Effectively removes Metadata acks & failure notices (see #1811) via no longer emitting the messages but does not fully remove them. They still technically exist, I'll remove them as part of a separate PR
- Completely reworks `CromwellApiServiceSpec` to actually be testing `CromwellApiService` and not a general integration test of our REST endpoints. Two specific tests didn't make the cut (#1828 and #1829) I'll address in separate PRs. There were other tests which did not make the cut but were already effectively being tested in their appropriate units.

### CODE REVIEW

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala [hunk 1]
```diff
@@ -15,7 +15,7 @@ trait MetadataSlickDatabase extends MetadataSqlDatabase {
 
   override def addMetadataEntries(metadataEntries: Iterable[MetadataEntry])
                                  (implicit ec: ExecutionContext): Future[Unit] = {
-    val action = dataAccess.metadataEntryIdsAutoInc ++= metadataEntries
+    val action = DBIO.seq(metadataEntries.grouped(2000).map(dataAccess.metadataEntries ++= _).toSeq:_*)
```
Comment (geoffjentry) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: I had forgotten about this. 2000 came from Rawls and the number was believed to have been Bernick wisdom. I'd like to make this configurable in a separate PR (this thing has already spidered enough as it is)

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreEngineActor.scala [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+import WorkflowStoreEngineActor._
+import cromwell.core.WorkflowId
+import cromwell.engine.workflow.workflowstore.WorkflowStoreActor._
+import cromwell.services.metadata.MetadataService.MetadataPutAcknowledgement
+import org.apache.commons.lang3.exception.ExceptionUtils
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+final case class WorkflowStoreEngineActor(store: WorkflowStore, serviceRegistryActor: ActorRef)
```
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreEngineActor.scala]: This is mostly the same as the original `WorkflowStoreActor`, I forget the exact changes but it was mostly to make the wiring work.

File: engine/src/main/scala/cromwell/webservice/CromwellApiService.scala [hunk 3]
```diff
@@ -53,14 +53,19 @@ trait CromwellApiService extends HttpService with PerRequestCreator {
       perRequest(requestContext, metadataBuilderProps, message)
   }
 
-  private def failBadRequest(t: Throwable, statusCode: StatusCode = StatusCodes.BadRequest) = respondWithMediaType(`application/json`) {
+  def handleQueryMetadataRequest(parameters: Seq[(String, String)]): Route = {
```
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/webservice/CromwellApiService.scala]: this made me cry a bit. I needed to override `handleMetadataRequest` in order to not actually have requests going into Real Cromwell. The query stuff worked a bit different and I wasn't able to figure out enough spray-fu to avoid just doing this.

**Comment by mcovarr** [comment]
:+1:

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,12 +6,16 @@
...
+## 25 
+
+* For MySQL users, a scalability improvement via batched DB writing of internal metadata events. **IMPORTANT** one must add `rewriteBatchedStatements=true` to their JDBC URL in their config in order to take advantage of this
 
 ## 24
 
 * When emitting workflow outputs to the Cromwell log only the first 1000 characters per output will be printed
 * Added support for conditional (`if`) statements.
 * Globs for Shared File System (SFS) backends, such as local or SGE, now use bash globbing instead of Java globbing, consistent with the JES backend.
+* For MySQL users, a massive scalability improvement via batched DB writing of internal metadata events. Note that one must add `rewriteBatchedStatements=true` to their JDBC URL in their config in order to take advantage of this
```
Comment (geoffjentry) [inline CHANGELOG.md]: Bad rebase on my part, most of this work predated 24's release

File: core/src/main/resources/reference.conf [hunk 5]
```diff
@@ -392,7 +392,25 @@ services {
   MetadataService {
     class = "cromwell.services.metadata.impl.MetadataServiceActor"
     # Set this value to "Inf" to turn off metadata summary refresh.  The default value is currently "2 seconds".
+    # FIXME: move
```
Comment (geoffjentry) [inline core/src/main/resources/reference.conf]: no, in fact i opened #1835 instead of just doing it here

File: core/src/main/resources/reference.conf [hunk 6]
```diff
@@ -392,7 +392,25 @@ services {
...
     # metadata-summary-refresh-interval = "Inf"
+    #
+    # config {
+      # For higher scale environments, e.g. many workflows and/or jobs, DB write performance for metadata events
+      # can improved by writing to the database in batches. Increasing this value can dramatically improve overall
+      # performance but will both lead to a higher memory usage as well as increase the risk that metadata events
+      # might not have been persisted in the event of a Cromwell crash.
+      #
+      # For normal usage the default value of 1 (effectively no batching) should be fine but for larger/production
+      # environments we recommend a value of at least 500. There'll be no one size fits all number here so testing
```
Comment (geoffjentry) [inline core/src/main/resources/reference.conf]: What about benchmarking? or actually "benchmarking & tuning" or something like that, i.e. making it clear that one should have concrete data to inform the tuning

File: core/src/main/resources/reference.conf [hunk 7]
```diff
@@ -392,7 +392,25 @@ services {
...
+      # is recommended.
+      # db-batch-size = 1
```
Comment (geoffjentry) [inline core/src/main/resources/reference.conf]: I originally did, in retrospect I think that's wrong for most users and dropped it back to 1. I don't think the defaults should be catering to a larger scale deployment and turning that number up does bring along some risk of loss of data in the event of a crash. Most small scale users will never notice the difference with the setting at 1, IMO that should be the default.

File: services/src/main/scala/cromwell/services/metadata/impl/WriteMetadataActor.scala [hunk 8]
```diff
@@ -1,29 +1,96 @@
...
+    case Event(ScheduledFlushToDb, curData) =>
+      log.debug("Initiating periodic metadata flush to DB")
+      goto(WritingToDb) using curData
+  }
+
+  when(WritingToDb) {
+    case Event(ScheduledFlushToDb, curData) => stay using curData
+    case Event(PutMetadataAction(events), curData) => stay using curData.addEvents(events)
+    case Event(FlushBatchToDb, curData) =>
+      curData match {
```
Comment (geoffjentry) [inline services/src/main/scala/cromwell/services/metadata/impl/WriteMetadataActor.scala]: Looking at this I don't particularly care for it either but I don't understand what you're after w/ that comment

File: core/src/main/resources/reference.conf (hunk 6 above)
Comment (mcovarr) [inline core/src/main/resources/reference.conf]: :1st_place_medal:

File: services/src/main/scala/cromwell/services/metadata/impl/WriteMetadataActor.scala (hunk 8 above)
Comment (mcovarr) [inline services/src/main/scala/cromwell/services/metadata/impl/WriteMetadataActor.scala]: Assuming the ADT:

```
case Event(FlushBatchToDb, NoEvents) => ...
case Event(FlushBatchToDb, SomeEvents(events)) => ...
```

Comment (geoffjentry) [inline services/src/main/scala/cromwell/services/metadata/impl/WriteMetadataActor.scala]: Oh right. Yes, that's why I didn't particularly care for it. Now that I reread I had misread `cases` as `classes` which made no sense (for good reason)

File: core/src/main/resources/reference.conf (hunk 5 above)
Comment (geoffjentry) [inline core/src/main/resources/reference.conf]: Phew I'm not going crazy. I thought I had fixed this and the changelog previously, I just hadn't pushed it.

File: core/src/main/resources/reference.conf [hunk 9]
```diff
@@ -409,7 +426,7 @@ database {
   #driver = "slick.driver.MySQLDriver$"
   #db {
   #  driver = "com.mysql.jdbc.Driver"
-  #  url = "jdbc:mysql://host/cromwell"
+  #  url = "jdbc:mysql://host/cromwell?rewriteBatchedStatements=true"
```
Comment (cjllanwarne) [inline core/src/main/resources/reference.conf]: This is an odd default considering the default batch size is `1`. Not saying the default is right or wrong one way or the other, but they're currently not really consistent with each other...

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala (hunk 1 above)
Comment (cjllanwarne) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: Add the ticket before merging and I'll let it pass :)

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreActor.scala [hunk 10]
```diff
@@ -1,242 +1,33 @@
...
-
-      MetadataEvent(MetadataKey(id, None, WorkflowMetadataKeys.SubmissionSection, WorkflowMetadataKeys.SubmissionSection_Workflow), MetadataValue(sourceFiles.wdlSource)),
-      MetadataEvent(MetadataKey(id, None, WorkflowMetadataKeys.SubmissionSection, WorkflowMetadataKeys.SubmissionSection_Inputs), MetadataValue(sourceFiles.inputsJson)),
-      MetadataEvent(MetadataKey(id, None, WorkflowMetadataKeys.SubmissionSection, WorkflowMetadataKeys.SubmissionSection_Options), MetadataValue(sourceFiles.workflowOptionsJson))
-    )
-
-    serviceRegistryActor ! PutMetadataAction(submissionEvents)
+  override def receive = {
+    case cmd: WorkflowStoreActorSubmitCommand => workflowStoreSubmitActor forward cmd
+    case cmd: WorkflowStoreActorEngineCommand => workflowStoreEngineActor forward cmd
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreActor.scala]: ToL (presumably a hangover from the single-actor days), these actors and their commands have `Submit` and `Engine` in different places.

Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreActor.scala]: Also ToL, `WorkflowStoreEngineActor` isn't a great name IMO (something something glass houses)... to me it should be either `EngineWorkflowStoreActor` or `WorkflowStoreManagerActor`

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreActor.scala [hunk 11]
```diff
@@ -1,242 +1,33 @@
...
-  when(Working) {
-    case Event(WorkDone, data) =>
-      val newData = data.pop
-      newData.currentOperation match {
-        case None => goto(Idle) using newData
-        case Some(WorkflowStoreActorCommandWithSender(cmd, sndr)) => startNewWork(cmd, sndr, newData)
-      }
-    case Event(cmd: WorkflowStoreActorCommand, data) => stay using data.withPendingCommand(cmd, sender)
-  }
+final case class WorkflowStoreActor(store: WorkflowStore, serviceRegistryActor: ActorRef) extends Actor with ActorLogging {
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreActor.scala]: Is there any benefit to keeping this redirection shell (other than not having to rewire everything)? It seems to me that we have two totally unrelated actors now that should stand alone.

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreEngineActor.scala [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+      val newData = data.pop
+      newData.currentOperation match {
+        case None => goto(Idle) using newData
+        case Some(WorkflowStoreActorCommandWithSender(cmd, sndr)) => startNewWork(cmd, sndr, newData)
+      }
+    case Event(cmd: WorkflowStoreActorEngineCommand, data) => stay using data.withPendingCommand(cmd, sender)
+  }
+
+  whenUnhandled {
+    case Event(MetadataPutAcknowledgement(_), _) =>
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreEngineActor.scala]: I thought this was gone now? Rebase needed?

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreSubmitActor.scala [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    * Runs processing on workflow source files before they are stored.
+    *
+    * @param processOptions How to process the workflow options
+    * @param source         Original workflow source
+    * @return Attempted updated workflow source
+    */
+  private def processSource(processOptions: WorkflowOptions => WorkflowOptionsJson)
+                           (source: WorkflowSourceFilesCollection): Future[WorkflowSourceFilesCollection] = {
+    val options = Future {
+      WorkflowOptions.fromJsonString(source.workflowOptionsJson)
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreSubmitActor.scala]: TOL: Presumably carry-over but I didn't think we even validated that it was JSON until the WFMaterializationActor, so that we get an ID and a record for every submission, even if it's invalid?

File: engine/src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala [hunk 14]
```diff
@@ -460,37 +311,23 @@ class CromwellApiServiceSpec extends FlatSpec with CromwellApiService with Scala
...
+     Get(s"/workflows/$version/${MockApiService.ExistingWorkflowId}/metadata?excludeKey=testKey2b&excludeKey=testKey3") ~>
+        cromwellApiService.metadataRoute ~>
+        check {
+          status should be(StatusCodes.OK)
+          val result = responseAs[JsObject]
+          result.fields.keys should contain allOf("testKey1a", "testKey1b", "testKey2a")
+          result.fields.keys should contain noneOf("testKey2b", "testKey3")
+          result.fields("testKey1a") should be(JsString("myValue1a"))
+          result.fields("testKey1b") should be(JsString("myValue1b"))
+          result.fields("testKey2a") should be(JsString("myValue2a"))
```
Comment (cjllanwarne) [inline engine/src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: ToL I'd rather have the test actor and the test expectations get these from a single known constant source, rather than having them be magical strings in both places.

File: services/src/main/scala/cromwell/services/metadata/impl/MetadataDatabaseAccess.scala [hunk 15]
```diff
@@ -78,10 +78,8 @@ trait MetadataDatabaseAccess {
       val value = metadataEvent.value map { _.value }
       val valueType = metadataEvent.value map { _.valueType.typeName }
       val jobKey = key.jobKey map { jk => (jk.callFqn, jk.index, jk.attempt) }
-      MetadataEntry(
-        workflowUuid, jobKey.map(_._1), jobKey.flatMap(_._2), jobKey.map(_._3), key.key, value, valueType, timestamp)
+      MetadataEntry(workflowUuid, jobKey.map(_._1), jobKey.flatMap(_._2), jobKey.map(_._3), key.key, value, valueType, timestamp)
```
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/metadata/impl/MetadataDatabaseAccess.scala]: ToL: Why do we bother exploding out the jobkey aboe? It seems like it then forces us to use the horrible `_._1` syntax on this line when we could instead say:
```
MetadataEntry(workflowUuid, key.jobKey.map(_.callFqn), key.jobKey.flatMap(_.index), key.jobKey.map(_.attempt), key.key, value, valueType, timestamp)
```

File: services/src/main/scala/cromwell/services/metadata/impl/WriteMetadataActor.scala [hunk 16]
```diff
@@ -1,29 +1,96 @@
...
-          val msg = MetadataPutFailed(action, t)
-          log.error(t, "Sending {} failure message {}", sndr, msg)
-          sndr ! msg
+  sealed trait WriteMetadataActorData {
+    def addEvents(newEvents: Iterable[MetadataEvent]): WriteMetadataActorData = {
+      if (newEvents.isEmpty) this
+      else {
+        val newEventsVector = newEvents.toVector
+        val combinedEvents: Vector[MetadataEvent] = this match {
+          case NoEvents => newEventsVector
```
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/metadata/impl/WriteMetadataActor.scala]: FWIW I like the theory here, but only if `HasEvents` enforces a non-empty value. Otherwise we aren't really using the type safety.
Similarly, if `newEvents` was types as non-empty, we shouldn't need the extra `if/else` checks here

**Comment by cjllanwarne** [comment]
Some genuine test failures and a lot of ToLs. Once you sort that all out, 👍 from me.

File: core/src/main/resources/reference.conf (hunk 9 above)
Comment (geoffjentry) [inline core/src/main/resources/reference.conf]: Disagree on inconsistency. This setting is simply saying if anything is intended to be a batch insert to allow it. The other setting is how we batch one particular setting.

More importantly however I wanted to make it as hard as possible for someone to not accidentally leave this off (it doesn't hurt anything AFAIK to leave it on), considering how many people just copy/paste this stuff that's likely.

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala (hunk 1 above)
Comment (geoffjentry) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: #1867

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreActor.scala (hunk 11 above)
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreActor.scala]: IMO it's the same concept as an interface where I changed the implementation behind the scenes. I disagree that they're wholly unrelated, together they form what we currently think of as the workflow store, whereas previously that was accomplished by a single point.

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreActor.scala (hunk 10 above)
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreActor.scala]: I think the different places was for an annoying but real reason. I'll look and see if I can recreate my thinking.

I think the latter suggestion is far too implementation specific in terms of exactly what is reading. I could be sold on the former (but don't see the big difference). Will also see if I can think of a different name. I wanted to go with e.g. read & write but the problem is the engine side also writes, just in a different manner

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala (hunk 1 above)
Comment (geoffjentry) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: Already existed as #1851

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreEngineActor.scala (hunk 12 above)
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreEngineActor.scala]: covered by the now poorly named #1811 - I had a comment somewhere in here (perhaps now gone) that this PR effectively removes these but doesn't remove stuff like this

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreSubmitActor.scala (hunk 13 above)
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreSubmitActor.scala]: I'll double check that I didn't hose something up here but I think the only thing I changed around here was turning it to being Future based. If I didn't in fact hose it up then I have no idea what was intended to happen here or what should :)

File: services/src/main/scala/cromwell/services/metadata/impl/MetadataDatabaseAccess.scala (hunk 15 above)
Comment (geoffjentry) [inline services/src/main/scala/cromwell/services/metadata/impl/MetadataDatabaseAccess.scala]: insert shrug emoji

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreActor.scala (hunk 10 above)
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreActor.scala]: Thinking about it more I think read & write *is* the right thing to do here now and the need to separate submit/engine is an artifact of something which no longer the case.

That does raise the question of the dispatcher but at the moment the way our dispatchers are configured i don't think it would matter anyways. I could also define a separate workflow store dispatcher

File: services/src/main/scala/cromwell/services/metadata/impl/WriteMetadataActor.scala (hunk 16 above)
Comment (geoffjentry) [inline services/src/main/scala/cromwell/services/metadata/impl/WriteMetadataActor.scala]: re your second sentence, admittedly it's not quite the same thing, but to me that's sort of like saying that all of the examples on how to avoid stringly typed code via wrapping them in case classes isn't helpful for extra type safety when that concept is the canonical example for such things. The idea there is that you might flub up on data ingest but you at least know you're plugging the right gozoutas into the right gozintas throughout your code.

I dont recall if/why I didn't consider moving the NEL concept further out to the outer edges in our code. Early on I was trying to avoid changing things believing this would be a surgical strike. Later on I was trying to avoid changing things as I was tired of changing things ;) The most likely answer is I encountered this in one of those two time periods.

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreSubmitActor.scala (hunk 13 above)
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreSubmitActor.scala]: Confirmed that it wasn't a me-change. You raise a good point, and I don't think this is working the way we intended. If I'm reading things right something which fails here won't return back and they'll get a timeout. 

I'll add a ticket to investigate

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreActor.scala (hunk 10 above)
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreActor.scala]: I realized that the reason I didn't do read & write didn't go away ;) Even the read operations involve some writing.

File: services/src/main/scala/cromwell/services/metadata/impl/WriteMetadataActor.scala [hunk 17]
```diff
@@ -1,29 +1,81 @@
...
+      }
+    case Event(ScheduledFlushToDb, curData) =>
+      log.debug("Initiating periodic metadata flush to DB")
+      goto(WritingToDb) using curData
+  }
+
+  when(WritingToDb) {
+    case Event(ScheduledFlushToDb, curData) => stay using curData
+    case Event(PutMetadataAction(events), curData) =>
+      val newData = curData ++ events.toVector
```
Comment (geoffjentry) [inline services/src/main/scala/cromwell/services/metadata/impl/WriteMetadataActor.scala]: Note to self: Cats provides a NonEmptyVector, investigate performance. ADT might be back in play
